shared/src/main/scala/mlscript/Typer.scala

```diff
@@ -62,15 +62,47 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool, val ne
       env: MutMap[Str, TypeInfo],
       mthEnv: MutMap[(Str, Str) \/ (Opt[Str], Str), MethodType],
       lvl: Int,
+      qenv: MutMap[Str, SkolemTag], // * SkolemTag for variables in quasiquotes
+      fvars: MutSet[ST], // * Free variables
+      inQuote: Bool, // * Is in quasiquote
       inPattern: Bool,
+      funDefs: MutMap[Str, DelayedTypeInfo],
       tyDefs: Map[Str, TypeDef],
       tyDefs2: MutMap[Str, DelayedTypeInfo],
       inRecursiveDef: Opt[Var], // TODO rm
       extrCtx: ExtrCtx,
   ) {
-    def +=(b: Str -> TypeInfo): Unit = env += b
+    def +=(b: Str -> TypeInfo): Unit = {
+      env += b
+      // * For lambdas and let bindings, we should not have bindings with the same name in one context.
+      // * This check (`!qenv.contains(b._1)`) is for UCS desugared terms that contain bindings and it is safe to share the same skolem tag
+      // * since the temporary variables are always bound.
+      if (inQuote && !qenv.contains(b._1)) {
+        val tag = SkolemTag(freshVar(NoProv, N, nameHint = S(b._1))(lvl))(NoProv)
+        println(s"Create skolem tag $tag for ${b._2} in quasiquote.")
+        qenv += b._1 -> tag
+      }
+    }
     def ++=(bs: IterableOnce[Str -> TypeInfo]): Unit = bs.iterator.foreach(+=)
-    def get(name: Str): Opt[TypeInfo] = env.get(name) orElse parent.dlof(_.get(name))(N)
+    def get(name: Str, quoted: Bool = false): Opt[TypeInfo] =
```

Comment:
Document what `quoted` means here and remove the error-prone default argument. Also, isn't there a better way? Why do we need this?

Reply:
I refactored the `get`-related code and documented the logic.